redwood-compose-testing/src/commonMain/kotlin/app/cash/redwood/compose/testing/WidgetValue.kt

```diff
@@ -15,10 +15,66 @@
  */
 package app.cash.redwood.compose.testing
 
+import app.cash.redwood.LayoutModifier
+
 /**
  * A widget that's implemented as a value class, appropriate for use in tests.
  *
  * Implementations of this interface may have lambda properties that trigger application behavior.
  * These lambda properties are **excluded** from [Any.equals], [Any.hashCode], and [Any.toString].
  */
-public interface WidgetValue
+public interface WidgetValue {
+  public val layoutModifiers: LayoutModifier
+
+  /** Returns all of the direct children of this widget, grouped by slot. */
+  public val childrenLists: List<List<WidgetValue>>
+    get() = listOf()
+}
+
+/**
+ * Returns a sequence that does a depth-first preorder traversal of the entire widget tree whose
+ * roots are this list. This is the same order elements occur in code.
+ *
+ * For example, given the following structure:
+ *
+ * ```
+ * Column {
+ *   Toolbar {
+ *     Icon(...)
+ *     Title(...)
+ *   }
+ *   Row {
+ *     Text(...)
+ *     Button(...)
+ *   }
+ * }
+ * ```
+ *
+ * The flattened elements are returned in this order:
+ *
+ * ```
+ * Column,
+ * Toolbar,
+ * Icon,
+ * Title,
+ * Row,
+ * Text,
+ * Button
+ * ```
+ */
+public fun List<WidgetValue>.flatten(): Sequence<WidgetValue> {
+  return sequence {
+    for (widget in this@flatten) {
+      flattenRecursive(widget)
+    }
+  }
```

Comment:
Not that it _really_ matters, but you could eliminate the separate loop with a `drop(1)`. Feel free to merge as-is.
```suggestion
  return sequence { flattenRecursive(this) }.drop(1) // Skip self.
```

Reply:
I don’t think it works 'cause `this` isn’t a WidgetValue; it’s a List.

Reply:
Oh, I missed the receiver. Wouldn't you have a single `WidgetValue` instance in the test to represent the content root?

Reply:
Totally unsure.

```
class RootValue  : WidgetValue {
  val children: List<WidgetValue>
}
```

vs.

```
val rootViews: List<WidgetValue>
```

We’d be boxing the list. If there’s something neat that lets us do, let’s do it.

Reply:
Yeah I guess the root is hidden and `snapshot()` is returning its contents which is almost always a single element, but someone _might_ do `setContent { Text(); Text() }` 🫤.

:shipit: as-is!